src/xnumber.ts

```diff
@@ -0,0 +1,188 @@
+  }
+
+  get inverse() {
+    if (this.den !== undefined) return new XNumber(this.den, this.num);
+    return new XNumber(1 / this.num, undefined, this.unit);
+  }
+
+  get negative() {
+    return new XNumber(-this.num, this.den, this.unit);
+  }
+
+  // ---------------------------------------------------------------------------
+
+  static fromString(s: string) {
+    // Replace whitespace and unit suffixes
+    s = s.replace(/\s/g, '').replace('–', '-');
+    const unit = s.endsWith('%') ? '%' : s.endsWith('π') ? 'π' : undefined;
+    if (unit) s = s.replace(unit, '');
+
+    // Handle integers or decimals
+    if (!s.includes('/')) return isNaN(+s) ? undefined : new XNumber(+s, undefined, unit);
+
+    // Handle fractions
+    const [num, den] = s.split('/').map(x => +x);
+    if (isNaN(num) || isNaN(den) || nearlyEquals(den, 0)) return;
+    if (!isInteger(num) || !isInteger(den)) return new XNumber(num / den, undefined, unit);
+    return new XNumber(num, den, unit);
+
+  }
+
+  static fractionFromDecimal(x: number, maxDen = 100) {
+    const sign = Math.sign(x);
+    const whole = Math.floor(Math.abs(x));
+    x = Math.abs(x) - whole;
+
+    let a = 0;
+    let b = 1;
+    let c = 1;
+    let d = 1;
+
+    while (b <= maxDen && d <= maxDen) {
+      const mediant = (a + c) / (b + d);
+      if (x === mediant) {
+        if (b + d <= maxDen) {
+          return new XNumber(sign * (whole * (b + d) + a + c), b + d);
+        } else if (d > b) {
+          return new XNumber(sign * (whole * d + c), d);
+        } else {
+          return new XNumber(sign * (whole * b + a), b);
+        }
+      } else if (x > mediant) {
+        [a, b] = [a + c, b + d];
+      } else {
+        [c, d] = [a + c, b + d];
+      }
+    }
+
+    if (b > maxDen) return new XNumber(sign * (whole * d + c), d);
+    return new XNumber(sign * (whole * b + a), b);
+  }
+
+  // ---------------------------------------------------------------------------
+
+  add(a: XNumber|number) {
+    return XNumber.sum(this, a);
+  }
+
+  subtract(a: XNumber|number) {
+    return XNumber.difference(this, a);
+  }
+
+  multiply(a: XNumber|number) {
+    return XNumber.product(this, a);
+  }
+
+  divide(a: XNumber|number) {
+    return XNumber.quotient(this, a);
+  }
+
+  /** Calculates the sum of two fractions a and b. */
+  static sum(a: XNumber, b: XNumber|number): XNumber {
+    if (typeof b === 'number') b = new XNumber(b);
+
+    // If units are different, always convert to a decimal
+    // TODO Maybe have special handling for fraction + percentage?
+    if (a.unit !== b.unit) return new XNumber(a.value + b.value);
+
+    // Neither a nor b are fractions
+    if (a.den === undefined && b.den === undefined) return new XNumber(a.num + b.num, undefined, a.unit);
+
+    // Ensure that a is always a fraction
+    if (a.den === undefined) [a, b] = [b, a];
+
+    // Trying to add a decimal to a fraction.
+    // TODO Maybe try XNumber.fractionFromDecimal?
```

Comment:
I don't know if this comment was written before `fractionFromDecimal` existed, but here's a reminder in case you did want to use it in this PR.